File: lib/services/database_service.dart
/// SQLite database service for local storage
import 'dart:convert';
import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';
import '../models/conversation.dart';

class DatabaseService {
  static Database? _database;
  static const String _dbName = 'omi_local.db';
  static const int _dbVersion = 3; // Incremented for tasks table

  static Future<Database> get database async {
    if (_database != null) return _database!;
    _database = await _initDatabase();
    return _database!;
  }

  static Future<Database> _initDatabase() async {
    final dbPath = await getDatabasesPath();
    final path = join(dbPath, _dbName);

    return await openDatabase(
      path,
      version: _dbVersion,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE conversations (
            id TEXT PRIMARY KEY,
            created_at INTEGER NOT NULL,
            title TEXT,
            summary TEXT,
            transcript TEXT
          )
        ''');
        
        await db.execute('''
          CREATE TABLE chat_messages (
            id TEXT PRIMARY KEY,
            conversation_id TEXT,
            text TEXT NOT NULL,
            is_user INTEGER NOT NULL,
            created_at INTEGER NOT NULL
          )
        ''');
        
        await db.execute('''
          CREATE TABLE memories (
            id TEXT PRIMARY KEY,
            content TEXT NOT NULL,
            category TEXT NOT NULL,
            created_at INTEGER NOT NULL,
            source_conversation_id TEXT
          )
        ''');
        
        await db.execute('''
          CREATE TABLE tasks (
            id TEXT PRIMARY KEY,
            title TEXT NOT NULL,
            description TEXT,
            due_date INTEGER,
            created_at INTEGER NOT NULL,
            source_conversation_id TEXT,
            is_completed INTEGER NOT NULL DEFAULT 0
          )
        ''');
      },
      onUpgrade: (db, oldVersion, newVersion) async {
        if (oldVersion < 2) {
          await db.execute('''
            CREATE TABLE IF NOT EXISTS memories (
              id TEXT PRIMARY KEY,
              content TEXT NOT NULL,
              category TEXT NOT NULL,
              created_at INTEGER NOT NULL,
              source_conversation_id TEXT
            )
          ''');
        }
        if (oldVersion < 3) {
          await db.execute('''
            CREATE TABLE IF NOT EXISTS tasks (
              id TEXT PRIMARY KEY,
              title TEXT NOT NULL,
              description TEXT,
              due_date INTEGER,
              created_at INTEGER NOT NULL,
              source_conversation_id TEXT,
              is_completed INTEGER NOT NULL DEFAULT 0
            )
          ''');
        }
      },
    );
  }

  // Conversation CRUD operations

  static Future<void> saveConversation(Conversation conversation) async {
    final db = await database;
    await db.insert(
      'conversations',
      {
        'id': conversation.id,
        'created_at': conversation.createdAt.millisecondsSinceEpoch,
        'title': conversation.title,
        'summary': conversation.summary,
        'transcript': jsonEncode(conversation.segments.map((s) => s.toJson()).toList()),
      },
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  static Future<List<Conversation>> getConversations({int limit = 50}) async {
    final db = await database;
    final rows = await db.query(
      'conversations',
      orderBy: 'created_at DESC',
      limit: limit,
    );
    return rows.map((row) => Conversation.fromDbRow(row)).toList();
  }

  static Future<Conversation?> getConversation(String id) async {
    final db = await database;
    final rows = await db.query(
      'conversations',
      where: 'id = ?',
      whereArgs: [id],
    );
    if (rows.isEmpty) return null;
    return Conversation.fromDbRow(rows.first);
  }

  static Future<void> deleteConversation(String id) async {
    final db = await database;
    await db.delete('conversations', where: 'id = ?', whereArgs: [id]);
  }

  static Future<String> getAllConversationsContext({int limit = 10}) async {
    final conversations = await getConversations(limit: limit);
    if (conversations.isEmpty) return '';
    
    return conversations.map((c) {
      return '''
--- Conversation from ${c.createdAt.toLocal()} ---
Title: ${c.title}
Summary: ${c.summary}
Transcript:
${c.transcript}
''';
    }).join('\n\n');
  }

  // Memory CRUD operations

  static Future<void> saveMemory(Memory memory) async {
    final db = await database;
    await db.insert(
      'memories',
      {
        'id': memory.id,
        'content': memory.content,
        'category': memory.category,
        'created_at': memory.createdAt.millisecondsSinceEpoch,
        'source_conversation_id': memory.sourceConversationId,
      },
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  static Future<List<Memory>> getMemories({int limit = 100}) async {
    final db = await database;
    final rows = await db.query(
      'memories',
      orderBy: 'created_at DESC',
      limit: limit,
    );
    return rows.map((row) => Memory.fromDbRow(row)).toList();
  }

  static Future<void> deleteMemory(String id) async {
    final db = await database;
    await db.delete('memories', where: 'id = ?', whereArgs: [id]);
  }

  /// Update memory content
  static Future<void> updateMemory(String id, String content) async {
    final db = await database;
    await db.update(
      'memories',
      {'content': content},
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  /// Check if a similar memory already exists (for deduplication)
  static Future<bool> hasSimilarMemory(String content) async {
    final db = await database;
    final normalizedContent = content.toLowerCase().trim();
    
    final rows = await db.query('memories');
    for (final row in rows) {
      final existingContent = (row['content'] as String).toLowerCase().trim();
      // Check for high similarity (simple substring or exact match)
      if (existingContent == normalizedContent || 
          existingContent.contains(normalizedContent) ||
          normalizedContent.contains(existingContent)) {
        return true;
      }
    }
    return false;
  }

  // Task CRUD operations

  static Future<void> saveTask(Task task) async {
    final db = await database;
    await db.insert(
      'tasks',
      {
        'id': task.id,
        'title': task.title,
        'description': task.description,
        'due_date': task.dueDate?.millisecondsSinceEpoch,
        'created_at': task.createdAt.millisecondsSinceEpoch,
        'source_conversation_id': task.sourceConversationId,
        'is_completed': task.isCompleted ? 1 : 0,
      },
      conflictAlgorithm: ConflictAlgorithm.replace,
    );
  }

  static Future<List<Task>> getTasks({int limit = 100}) async {
    final db = await database;
    final rows = await db.query(
      'tasks',
      orderBy: 'is_completed ASC, due_date ASC, created_at DESC',
      limit: limit,
    );
    return rows.map((row) => Task.fromDbRow(row)).toList();
  }

  static Future<void> updateTaskCompletion(String id, bool isCompleted) async {
    final db = await database;
    await db.update(
      'tasks',
      {'is_completed': isCompleted ? 1 : 0},
      where: 'id = ?',
      whereArgs: [id],
    );
  }

  static Future<void> deleteTask(String id) async {
    final db = await database;
    await db.delete('tasks', where: 'id = ?', whereArgs: [id]);
  }

  /// Check if a similar task already exists (for deduplication)
  static Future<bool> hasSimilarTask(String title) async {
    final db = await database;
    final normalizedTitle = title.toLowerCase().trim();
    
    final rows = await db.query('tasks', where: 'is_completed = 0');
    for (final row in rows) {
      final existingTitle = (row['title'] as String).toLowerCase().trim();
      if (existingTitle == normalizedTitle || 
          existingTitle.contains(normalizedTitle) ||
          normalizedTitle.contains(existingTitle)) {
        return true;
      }
    }
    return false;
  }

  /// Export all data for backup/sharing
  static Future<Map<String, dynamic>> exportAllData() async {
    final conversations = await getConversations(limit: 10000);
    final memories = await getMemories(limit: 10000);
    final tasks = await getTasks(limit: 10000);
    
    return {
      'export_date': DateTime.now().toIso8601String(),
      'app_version': '2.1.0',
      'conversations': conversations.map((c) => c.toJson()).toList(),
      'memories': memories.map((m) => m.toJson()).toList(),
      'tasks': tasks.map((t) => t.toJson()).toList(),
    };
  }
}

File: OmiLocal/Sources/Services/DatabaseService.swift
import Foundation
import SQLite

class DatabaseService {
    static let shared = DatabaseService()
    
    private var db: Connection?
    
    private let conversations = Table("conversations")
    private let memories = Table("memories")
    private let tasks = Table("tasks")
    
    private let id = SQLite.Expression<String>("id")
    private let createdAt = SQLite.Expression<Int64>("created_at")
    private let title = SQLite.Expression<String?>("title")
    private let summary = SQLite.Expression<String?>("summary")
    private let transcript = SQLite.Expression<String?>("transcript")
    private let content = SQLite.Expression<String>("content")
    private let category = SQLite.Expression<String>("category")
    private let sourceConversationId = SQLite.Expression<String?>("source_conversation_id")
    private let descriptionCol = SQLite.Expression<String?>("description")
    private let dueDate = SQLite.Expression<Int64?>("due_date")
    private let isCompleted = SQLite.Expression<Int64>("is_completed")
    
    init() {}
    
    func initialize() async {
        let fileManager = FileManager.default
        let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dbPath = documentsURL.appendingPathComponent("omi_local.db").path
        
        do {
            db = try Connection(dbPath)
            
            try db?.run(conversations.create(ifNotExists: true) { t in
                t.column(id, primaryKey: true)
                t.column(createdAt)
                t.column(title)
                t.column(summary)
                t.column(transcript)
            })
            
            try db?.run(memories.create(ifNotExists: true) { t in
                t.column(id, primaryKey: true)
                t.column(content)
                t.column(category)
                t.column(createdAt)
                t.column(sourceConversationId)
            })
            
            try db?.run(tasks.create(ifNotExists: true) { t in
                t.column(id, primaryKey: true)
                t.column(title)
                t.column(descriptionCol)
                t.column(dueDate)
                t.column(createdAt)
                t.column(sourceConversationId)
                t.column(isCompleted, defaultValue: 0)
            })
            
            print("Database initialized at: \(dbPath)")
        } catch {
            print("Database initialization failed: \(error)")
        }
    }
    
    func saveConversation(_ conversation: Conversation) async {
        guard let db = db else { return }
        
        let transcriptJson = try? JSONEncoder().encode(conversation.segments)
        let transcriptStr = transcriptJson.flatMap { String(data: $0, encoding: .utf8) }
        
        do {
            let insert = conversations.insert(or: .replace,
                id <- conversation.id,
                createdAt <- Int64(conversation.createdAt.timeIntervalSince1970),
                title <- conversation.title,
                summary <- conversation.summary,
                transcript <- transcriptStr
            )
            try db.run(insert)
        } catch {
            print("Failed to save conversation: \(error)")
        }
    }
    
    func getConversations(limit: Int = 50) async -> [Conversation] {
        guard let db = db else { return [] }
        
        var result: [Conversation] = []
        
        do {
            for row in try db.prepare(conversations.order(createdAt.desc).limit(limit)) {
                var segments: [TranscriptSegment] = []
                if let transcriptStr = row[transcript],
                   let data = transcriptStr.data(using: .utf8) {
                    segments = (try? JSONDecoder().decode([TranscriptSegment].self, from: data)) ?? []
                }
                
                let conversation = Conversation(
                    id: row[id],
                    createdAt: Date(timeIntervalSince1970: TimeInterval(row[createdAt])),
                    title: row[title] ?? "",
                    summary: row[summary] ?? "",
                    segments: segments
                )
                result.append(conversation)
            }
        } catch {
            print("Failed to get conversations: \(error)")
        }
        
        return result
    }
    
    func deleteConversation(_ convId: String) async {
        guard let db = db else { return }
        
        do {
            let conversation = conversations.filter(id == convId)
            try db.run(conversation.delete())
        } catch {
            print("Failed to delete conversation: \(error)")
        }
    }
    
    func saveMemory(_ memory: Memory) async {
        guard let db = db else { return }
        
        do {
            let insert = memories.insert(or: .replace,
                id <- memory.id,
                content <- memory.content,
                category <- memory.category,
                createdAt <- Int64(memory.createdAt.timeIntervalSince1970),
                sourceConversationId <- memory.sourceConversationId
            )
            try db.run(insert)
        } catch {
            print("Failed to save memory: \(error)")
        }
    }
    
    func getMemories(limit: Int = 100) async -> [Memory] {
        guard let db = db else { return [] }
        
        var result: [Memory] = []
        
        do {
            for row in try db.prepare(memories.order(createdAt.desc).limit(limit)) {
                let memory = Memory(
                    id: row[id],
                    content: row[content],
                    category: row[category],
                    createdAt: Date(timeIntervalSince1970: TimeInterval(row[createdAt])),
                    sourceConversationId: row[sourceConversationId]
                )
                result.append(memory)
            }
        } catch {
            print("Failed to get memories: \(error)")
        }
        
        return result
    }
    
    func deleteMemory(_ memId: String) async {
        guard let db = db else { return }
        
        do {
            let memory = memories.filter(id == memId)
            try db.run(memory.delete())
        } catch {
            print("Failed to delete memory: \(error)")
        }
    }
    
    func hasSimilarMemory(_ memContent: String) async -> Bool {
        guard let db = db else { return false }
        
        do {
            let query = memories.filter(content.like("%\(memContent)%"))
            return try db.scalar(query.count) > 0
        } catch {
            return false
        }
    }
    
    func updateMemory(_ memId: String, newContent: String) async {
        guard let db = db else { return }
        
        do {
            let memory = memories.filter(id == memId)
            try db.run(memory.update(content <- newContent))
        } catch {
            print("Failed to update memory: \(error)")
        }
    }
    
    func saveTask(_ task: TodoItem) async {
        guard let db = db else { return }
        
        do {
            let insert = tasks.insert(or: .replace,
                id <- task.id,
                title <- task.title,
                descriptionCol <- task.description,
                dueDate <- task.dueDate.map { Int64($0.timeIntervalSince1970) },
                createdAt <- Int64(task.createdAt.timeIntervalSince1970),
                sourceConversationId <- task.sourceConversationId,
                isCompleted <- task.isCompleted ? 1 : 0
            )
            try db.run(insert)
        } catch {
            print("Failed to save task: \(error)")
        }
    }
    
    func getTasks(limit: Int = 100) async -> [TodoItem] {
        guard let db = db else { return [] }
        
        var result: [TodoItem] = []
        
        do {
            for row in try db.prepare(tasks.order(createdAt.desc).limit(limit)) {
                let task = TodoItem(
                    id: row[id],
                    title: row[title] ?? "",
                    description: row[descriptionCol],
                    dueDate: row[dueDate].map { Date(timeIntervalSince1970: TimeInterval($0)) },
                    createdAt: Date(timeIntervalSince1970: TimeInterval(row[createdAt])),
                    sourceConversationId: row[sourceConversationId],
                    isCompleted: row[isCompleted] == 1
                )
                result.append(task)
            }
        } catch {
            print("Failed to get tasks: \(error)")
        }
        
        return result
    }
    
    func deleteTask(_ taskId: String) async {
        guard let db = db else { return }
        
        do {
            let task = tasks.filter(id == taskId)
            try db.run(task.delete())
        } catch {
            print("Failed to delete task: \(error)")
        }
    }
    
    func updateTaskCompletion(taskId: String, completed: Bool) async {
        guard let db = db else { return }
        
        do {
            let task = tasks.filter(id == taskId)
            try db.run(task.update(isCompleted <- (completed ? 1 : 0)))
        } catch {
            print("Failed to update task: \(error)")
        }
    }
    
    func hasSimilarTask(_ taskTitle: String) async -> Bool {
        guard let db = db else { return false }
        
        do {
            let query = tasks.filter(title.like("%\(taskTitle)%"))
            return try db.scalar(query.count) > 0
        } catch {
            return false
        }
    }
    
    func exportAllData() async -> [String: Any] {
        let convs = await getConversations(limit: 1000)
        let mems = await getMemories(limit: 1000)
        let tks = await getTasks(limit: 1000)
        
        return [
            "conversations": convs.map { conv in
                [
                    "id": conv.id,
                    "created_at": conv.createdAt.timeIntervalSince1970,
                    "title": conv.title,
                    "summary": conv.summary,
                    "segments": conv.segments.map { $0.toDictionary() }
                ]
            },
            "memories": mems.map { mem in
                [
                    "id": mem.id,
                    "content": mem.content,
                    "category": mem.category,
                    "created_at": mem.createdAt.timeIntervalSince1970,
                    "source_conversation_id": mem.sourceConversationId ?? ""
                ]
            },
            "tasks": tks.map { tk in
                [
                    "id": tk.id,
                    "title": tk.title,
                    "description": tk.description ?? "",
                    "due_date": tk.dueDate?.timeIntervalSince1970 ?? 0,
                    "created_at": tk.createdAt.timeIntervalSince1970,
                    "is_completed": tk.isCompleted
                ]
            },
            "exported_at": Date().timeIntervalSince1970
        ]
    }
}

extension TranscriptSegment {
    func toDictionary() -> [String: Any] {
        return [
            "text": text,
            "speaker_id": speakerId,
            "start": startTime,
            "end": endTime,
            "is_user": isUser
        ]
    }
}
